### PR TITLE
First making model.py (Generation Model : CVAE)

### PR DESCRIPTION
**Pull request by hsmtknj**
make model.py (Conditional Variational AutoEncoder)

[References]
・猫でもわかるVariational AutoEncoder
https://www.slideshare.net/ssusere55c63/variational-autoencoder-64515581

・kerasのmnistのサンプルを読んでみる
https://qiita.com/ash8h/items/29e24fc617b832fba136

・Kerasはfunctional APIもきちんと理解しよう
https://qiita.com/shu_marubo/items/eb297b2245040e50c00f

### CODE REVIEW

File: cvae/model.py [hunk 1]
```diff
@@ -29,11 +32,11 @@ def _sampling(self, args):
         """
         mean, log_sigma = args
 
-        #TODO: generating random variable epsilon ~ N(0, I)
-        epsilon = 0
+        # TODO: generating random variable epsilon ~ N(0, I)
+        epsilon = np.random.normal(0, 1, z_dim)
```
Comment (hstko) [inline cvae/model.py]: 新しく自分で特殊なレイヤーを定義するのは若干難しかったから説明不足で申し訳ないです．
とりあえず先に直し方だけ書いておきます！
詳しくは下に書いておきます．
epsilon = K.random_normal(shape=(K.shape(z_mean)[0], self.z_dim), mean=0.,stddev=1.0)

ちゃんとfunctionのコメントで引数の型を詳しく書いてなかったから申し訳なかったんやけど，keras は基本データは実際のfloat とかの数値とかではなくてテンソルのオブジェクトになってるんよね！だから，演算するときはkeras のbackendモジュールのfunctionを使って演算をしないといけない！
これは多分なんやけど，kerasがそもそも tensorflow とかのラッパーで，tensorflowとかは基本的に先に計算グラフ定義して，あとで実際のデータを流すってとこに関係してる気がする！
だからsumとかexpとか乱数とかshapeとかだいたいbackendモジュールを使ってやる必要があるんよね．
本来，こういう実際の演算部分はkerasのデベロッパーが書いてくれてて，そんな複雑ではない Deep Leaning 試すだけならユーザーは意識することないんやけど，今回はサンプリングレイヤーっていう特殊な層を定義する必要があったからこうなってるんよな！

Comment (hsmtknj) [reply cvae/model.py]: 基本的にKerasの入力とか出力はtensorflowのオブジェクトで，概念的にしか記述できひん みたいなこと？よって，演算系もその入出力のオブジェクトに合わせた特殊な形で記述せなあかんみたいな理解であってる？

今まで基本的に引数には実際のintとかfloatとかを入れて関数書いてたから，ちょっとなれる必要ありそうやな．

train.pyとかでは実際に生のデータを引数として渡すけど，その値は内部的になにかしらの処理（tensorflowの特殊なオブジェクト化的な？）がされて，このmodel.pyに特殊なオブジェクトの形で引数として渡されるんかな？

修正箇所はこのepsilonの箇所だけでいいんかな？
修正したら再度プルリクなげます！

Comment (hsmtknj) [reply cvae/model.py]: というか，あれか．たくも書いてくれてるけど，ニューラルネットの内部の実装を直接いじりたいときは要注意ってことか．

File: cvae/model.py [hunk 2]
```diff
@@ -29,11 +32,11 @@ def _sampling(self, args):
         """
         mean, log_sigma = args
 
-        #TODO: generating random variable epsilon ~ N(0, I)
-        epsilon = 0
+        # TODO: generating random variable epsilon ~ N(0, I)
```
Comment (takami-sris) [inline cvae/model.py]: 実装ができたらTODOコメントは消しちゃってもいいと思いましたー。
（他のとこもネ）

File: cvae/model.py (hunk 1 above)
Comment (hstko) [reply cvae/model.py]: うーん，kerasの入出力はtensorflowのオブジェクトって訳でもない！tensorflowじゃなくてtheanoとかも
裏で読んだり出来るし基本的には入出力は実体のないもの的な感じかな・・・
f(X) = exp(2x) に対してlim(x->1) f(x) = 2e とか lim(x->2) f(x) = 2e^2 とかって計算するんじゃなくて，先に f'(x) = 2exp(x)って計算して，あとでデータx=[1, 2]とかを投げるイメージやと思う！ 
ごめん，俺も正確には理解してないし，概念の説明もむずかしい・・・
多分，またこれが落ち着いてから１回tensorflowの計算グラフを一緒に勉強した方がいい気がする！
まあこういうやつは tensorflowで書くときは慣れないとあかんのやけど，kerasで書く以上は普通はそんな慣れる必要もないんやけどね・・・
だからこそ基本的にはまずはkeras でざっと勉強して必要があればtensorflowとかtheanoとか使って慣れるとより良いんやと思う！

File: cvae/model.py [hunk 3]
```diff
@@ -29,11 +32,11 @@ def _sampling(self, args):
...
+        epsilon = np.random.normal(0, 1, z_dim)
 
-        #TODO: return random variable z. calculating z ~ N(mean, exp(log_sigma)) using z = mean + exp(log_sigma) * epsilon
-        return None
+        # TODO: return random variable z. calculating z ~ N(mean, exp(log_sigma)) using z = mean + exp(log_sigma) * epsilon
+        return mean + Math.exp(log_sigma) * epsilon
```
Comment (hstko) [inline cvae/model.py]: ここも K.exp() でお願いします！
多分numpy とかscipyとかにある基本的な演算とか関数はbackendでも用意されてる気がする！

File: cvae/model.py [hunk 4]
```diff
@@ -43,32 +46,32 @@ def build_cvae_mlp(self, kl_weight=1.0):
         """
         input_dim = self.input_dim
 
-        #TODO: defining x and y using Input Layer
-        x = None
-        y = None
+        # TODO: defining x and y using Input Layer
+        x = Input(shape=input_dim,)
```
Comment (hstko) [inline cvae/model.py]: Input(shape=(input_dim,))　かな！
ここではshapeは整数のタプルじゃないとあかんはず！
バッチサイズは入れなくていい感じで！

File: cvae/model.py [hunk 5]
```diff
@@ -43,32 +46,32 @@ def build_cvae_mlp(self, kl_weight=1.0):
...
         # TODO: concatenating x and y using Concatenate Layer
-        concat = None
+        input_layer = Concatenate([x, y])
 
         # TODO: building a encoder network
-        enc_dense = None #dense layer
-        enc_drop = None #dropout layer
-        z_mean = None #mean layer (dense layer)
-        z_log_sigma = None #variance layer(dense layer)
+        enc_dense = Dense(input_dim, activation='relu')(input_layer)  # dense layer 1
```
Comment (takami-sris) [inline cvae/model.py]: Denseの第一引数ってそのレイヤーの出力する次元数よね？
input_dimって変数名だとちょっと混乱するかもしれないねー。（あれ入力だっけ？みたいな。）

File: cvae/model.py [hunk 6]
```diff
@@ -43,32 +46,32 @@ def build_cvae_mlp(self, kl_weight=1.0):
...
         z = Lambda(self._sampling)([z_mean, z_log_sigma])
 
         # TODO: concatenating z and y using Concatenate Layer
-        dec_merged = None
+        dec_merged = Concatenate([z, y])
 
         # TODO: building a decoder network
-        dec_dense = None # dense layer
-        dec_out = None # output layer(dense layer) using sigmoid. you should use use_bias=False
+        dec_dense = Dense(input_dim, activation='relu')(dec_merged)  # dense layer
```
Comment (takami-sris) [inline cvae/model.py]: ここもinput_dimはなんか混乱しそう・・。
個人的にそう思っただけなんでもし二人がinput_dimでいいってなったらそれでOKです( ´ ▽ ` )

File: cvae/model.py [hunk 7]
```diff
@@ -43,32 +46,32 @@ def build_cvae_mlp(self, kl_weight=1.0):
         """
         input_dim = self.input_dim
 
-        #TODO: defining x and y using Input Layer
-        x = None
-        y = None
+        # TODO: defining x and y using Input Layer
+        x = Input(shape=input_dim,)
+        y = Input(shape=self.n_classes,)
```
Comment (hstko) [inline cvae/model.py]: 上に同じ！

File: cvae/model.py [hunk 8]
```diff
@@ -43,32 +46,32 @@ def build_cvae_mlp(self, kl_weight=1.0):
...
+        input_layer = Concatenate([x, y])
 
         # TODO: building a encoder network
-        enc_dense = None #dense layer
-        enc_drop = None #dropout layer
-        z_mean = None #mean layer (dense layer)
-        z_log_sigma = None #variance layer(dense layer)
+        enc_dense = Dense(input_dim, activation='relu')(input_layer)  # dense layer 1
+        # enc_drop = None #dropout layer
+        z_mean = Dense(self.z_dim, activation='relu')  # mean layer (dense layer)
```
Comment (hstko) [inline cvae/model.py]: Dense(self.z_dim, activation='relu')(enc_dense) で修正お願いします！
前の層（の出力）を今定義した層の引数（入力）にする感じで計算グラフを定義していくイメージかな！

File: cvae/model.py [hunk 9]
```diff
@@ -43,32 +46,32 @@ def build_cvae_mlp(self, kl_weight=1.0):
...
+        z_log_sigma = Dense(self.z_dim, activation='relu')  # variance layer (dense layer)
```
Comment (hstko) [inline cvae/model.py]: 上に同じで！

File: cvae/model.py [hunk 10]
```diff
@@ -43,32 +46,32 @@ def build_cvae_mlp(self, kl_weight=1.0):
...
 
         # TODO: concatenating z and y using Concatenate Layer
-        dec_merged = None
+        dec_merged = Concatenate([z, y])
 
         # TODO: building a decoder network
-        dec_dense = None # dense layer
-        dec_out = None # output layer(dense layer) using sigmoid. you should use use_bias=False
+        dec_dense = Dense(input_dim, activation='relu')(dec_merged)  # dense layer
+        dec_out = Dense(input_dim, activation='sigmoid')  # output layer(dense layer) using sigmoid. you should use use_bias=False
```
Comment (hstko) [inline cvae/model.py]: 上に同じ！Dense(input_dim, activation='sigmoid')(dec_dense)

File: cvae/model.py (hunk 2 above)
Comment (hsmtknj) [reply cvae/model.py]: 確かに，消しときます！

File: cvae/model.py (hunk 5 above)
Comment (hsmtknj) [reply cvae/model.py]: そこの次元数いくつに設定すればいいかわからんかったから，とりあえず一緒の次元にしときました．なんか適当な変数名に変更しとくわ！

**Review by hstko — CHANGES_REQUESTED** [review]
計算グラフを構築してからデータ流すイメージがなれないよねー・・・
あと，今回はどっちでもいいんだけれど，あとでネットワーク可視化したり，特定の層の重みをとってくる時は，layerを定義するところで，name=''で名前つけたら便利やで！
Ex)
Dense(input_dim, activation='sigmoid', name='output_layer')

File: cvae/model.py (hunk 5 above)
Comment (hstko) [reply cvae/model.py]: あ，これは違う！
self.intermediate_dim を使ってもらうつもりでした！

File: cvae/model.py (hunk 6 above)
Comment (hstko) [reply cvae/model.py]: ここも self.intermediate_dim

File: cvae/model.py (hunk 3 above)
Comment (hsmtknj) [reply cvae/model.py]: numpy.〜的なんは基本的にK.〜に置き換えて差し支えないってことかな？
修正しときます！

名前もつけとくわ！

File: cvae/model.py (hunk 4 above)
Comment (hsmtknj) [reply cvae/model.py]: got it !

**Comment by takami-sris** [comment]
@hstko ちょっと質問なんですが（多分作成したのはtakuなので）
```
build_cvae_mlp
```
これの先頭に"_"ついてないのはmodel.py以外でも使うからですかね？( ´ ▽ ` )

**Comment by takami-sris** [comment]
@hsmtknj あとでpull req見直すこともあるので、タイトルわかりやすいように良い感じにしといてくださいー！( ´ ▽ ` )

File: cvae/model.py (hunk 5 above)
Comment (hsmtknj) [reply cvae/model.py]: そうか，申し訳ない！
中間層は全部このself.intermediate_dimのサイズでええかな？

**Review by hsmtknj — COMMENTED** [review]
@takamisris タイトル，了解です〜

**Review by hstko — APPROVED** [review]
多分大丈夫だと思います！

**Review by takami-sris — APPROVED** [review]
